storage/kafka/datasource.go

```diff
@@ -0,0 +1,349 @@
+			}
+			batchSize, err := config.GetInt(dbConfig, "batchSize", config.WithDefault(1))
+			if err != nil {
+				return nil, errors.Wrap(err, "couldn't get batch size")
+			}
+			decodeAsJSON, err := config.GetBool(dbConfig, "json", config.WithDefault(false))
+			if err != nil {
+				return nil, errors.Wrap(err, "couldn't get json option")
+			}
+
+			brokers := make([]string, len(hosts))
+			for i := range hosts {
+				brokers[i] = fmt.Sprintf("%s:%d", hosts[i], ports[i])
+			}
+
+			return &DataSource{
+				brokers:      brokers,
+				topic:        topic,
+				partition:    partition,
+				batchSize:    batchSize,
+				decodeAsJSON: decodeAsJSON,
+				alias:        alias,
+				stateStorage: matCtx.Storage,
+			}, nil
+		},
+		nil,
+		availableFilters,
+		metadata.Unbounded,
+		partitions,
+	)
+}
+
+// NewDataSourceBuilderFactoryFromConfig creates a data source builder factory using the configuration.
+func NewDataSourceBuilderFactoryFromConfig(dbConfig map[string]interface{}) (physical.DataSourceBuilderFactory, error) {
+	partitions, err := config.GetInt(dbConfig, "partitions", config.WithDefault(1))
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't get primaryKeys")
+	}
+
+	return NewDataSourceBuilderFactory(partitions), nil
+}
+
+var offsetPrefix = []byte("$kafka_offset$")
+var tokenQueuePrefix = []byte("$token_queue$")
+
+func (ds *DataSource) Get(ctx context.Context, variables octosql.Variables, streamID *execution.StreamID) (execution.RecordStream, error) {
+	tx := storage.GetStateTransactionFromContext(ctx).WithPrefix(streamID.AsPrefix())
+
+	dialer := &kafka.Dialer{
+		Timeout:   10 * time.Second,
+		DualStack: true,
+	}
+
+	r := kafka.NewReader(kafka.ReaderConfig{
+		Brokers:   ds.brokers,
+		Topic:     ds.topic,
+		Dialer:    dialer,
+		Partition: ds.partition,
+		MinBytes:  10e1,
+		MaxBytes:  10e7,
+	})
+
+	rs := &RecordStream{
+		stateStorage: ds.stateStorage,
+		streamID:     streamID,
+		kafkaReader:  r,
+		batchSize:    ds.batchSize,
+		decodeAsJSON: ds.decodeAsJSON,
+		alias:        ds.alias,
+	}
+	err := rs.loadOffset(tx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "couldn't load kafka partition %d offset", ds.partition)
+	}
+
+	tokenQueue := execution.NewOutputQueue(tx.WithPrefix(tokenQueuePrefix))
+	for i := 0; i < 5; i++ {
+		token := octosql.MakePhantom()
+		err := tokenQueue.Push(ctx, &token)
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't push batch token to queue")
+		}
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	rs.workerCtxCancel = cancel
+	rs.workerCloseErrChan = make(chan error)
+
+	go func() {
+		rs.RunWorker(ctx)
+		log.Println("worker done")
+	}()
+
+	return rs, nil
+}
+
+type RecordStream struct {
+	stateStorage storage.Storage
+	streamID     *execution.StreamID
+	kafkaReader  *kafka.Reader
+	batchSize    int
+	decodeAsJSON bool
+	alias        string
+
+	workerCtxCancel    func()
+	workerCloseErrChan chan error
+}
+
+var outputQueuePrefix = []byte("$output_queue$")
```

Comment:
Can we keep all globals together so they are easier to find?

Reply:
I'm putting them before the first function they are used. I can change this, but it's like that everywhere so far.